### ISSUE TITLE
potential miscompile involving raw pointers

### ISSUE THREAD

**Issue by godmar** [rust-lang/rust#87352]
Upfront, I will admit that I'm new to Rust and very hesitant to file this report as it involves raw pointers and a small amount of unsafe code.  However, there may be a small chance that the behavior I'm seeing is not desired or is not invoking UB, or that if it is suffering from UB, the UB may be unrelated to the behavior.

Synopsis:
I'm observing a case where the compiler performs an optimization such that the optimized code exhibits different behavior than the unoptimized code. (The unoptimized code completes as expected, the optimized code enters an infinite loop because the compiler assumed that a particular memory location didn't change, even though it is accessed and changed via a raw pointer.)

In short, the loop in question is:
```rust
    fn remove(this: *mut ListElem) {
        unsafe {
            (*(*this).prev).next = (*this).next;
            (*(*this).next).prev = (*this).prev;
        }
    }
...
    fn pop_back(&self) {
        ListElem::remove(self.tail.prev);                                     
    }   
    
    #[inline(never)]                                                          
    fn drain(&self) {
        while self.head.next != ptr::addr_of!(self.tail) as *mut ListElem {   
            self.pop_back();
        }                                                                     
    }
```
If `pop_back()` is inlined (as it is under normal invocation with `-O`), then it will use the value of `self.tail.prev` as it was on entry to `drain()`, rather than reloading the value on each call to `pop_back()`. 
In other words, `drain()` is compiled as if it had been written as:
```rust
    fn drain(&self) {
        let tmp = self.tail.prev;
        while self.head.next != ptr::addr_of!(self.tail) as *mut ListElem {
            ListElem::remove(tmp);
        }
    }
```

I've verified this by examining the assembly code, and you may also add a `#[inline(never)]` to `pop_back()` and verify that the behavior under optimization changes.

[Playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=c8201a65f9ee32724f4af273a2a2b560) (to reproduce timeout, compile with 'release')

I tried this code (either via `rustc` or via `cargo`)

```rust
// testmin.rs
use std::ptr;
use std::boxed::Box;
use std::pin::Pin;

#[derive(Debug)]
struct ListElem {
    prev: *mut ListElem,
    next: *mut ListElem,
}

impl ListElem {
    fn new() -> ListElem {
        ListElem {
            prev: ptr::null_mut(),
            next: ptr::null_mut(),
        }
    }

    fn remove(this: *mut ListElem) {
        unsafe {
            (*(*this).prev).next = (*this).next;
            (*(*this).next).prev = (*this).prev;
        }
    }
}

struct List {
    head: ListElem,
    tail: ListElem,
}

impl <'a> List {
    fn new() -> Pin<Box<List>> {
        let mut list = Box::pin(List {
            head: ListElem::new(),
            tail: ListElem::new(),
        });
        list.head.next = ptr::addr_of!(list.tail) as *mut ListElem;
        list.tail.prev = ptr::addr_of!(list.head) as *mut ListElem;
        list
    }

    // if pop_back is inlined, drain is miscompiled because it assumes
    // that `self.tail.prev` doesn't need to be reloaded between calls to
    // self.pop_back()
    //#[inline(never)]
    fn pop_back(&self) {
        ListElem::remove(self.tail.prev);
    }

    #[inline(never)]
    fn drain(&self) {
        while self.head.next != ptr::addr_of!(self.tail) as *mut ListElem {
            self.pop_back();
        }
    }
}

fn main() {
    let mut l = List::new();

    println!("part 1, build a list with 2 elements");
    let mut e1 = Box::pin(ListElem::new());
    let ep1 : *mut ListElem = &mut *e1;

    let mut e2 = Box::pin(ListElem::new());
    let ep2 : *mut ListElem = &mut *e2;

    unsafe {
        // set forward pointers head -> e1 -> e2 -> tail
        l.head.next = ep1;
        (*ep1).next = ep2;
        (*ep2).next = ptr::addr_of!(l.tail) as *mut ListElem;
        // set backward pointers tail -> e2 -> e1 -> head
        l.tail.prev = ep2;
        (*ep2).prev = ep1;
        (*ep1).prev = ptr::addr_of!(l.head) as *mut ListElem;
    }

    println!("trying to drain list");
    l.drain();
    println!("complete");
}
```

I expected to see this happen: 
```
part 1, build a list with 2 elements
trying to drain list
complete
```

Instead, this happened: 
```
part 1, build a list with 2 elements
trying to drain list
```
(program enters infinite loop)

### Meta
This behavior occurs under 

```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-unknown-linux-gnu
release: 1.53.0
LLVM version: 12.0.1
```

as well as under nightly:
```
rustc 1.55.0-nightly (7a16cfcff 2021-07-11)
binary: rustc
commit-hash: 7a16cfcffc58cd6994984a2c014b0cc27ed8f66b
commit-date: 2021-07-11
host: x86_64-unknown-linux-gnu
release: 1.55.0-nightly
LLVM version: 12.0.1
```
It occurs _only_ under `-O` and it doesn't depend on whether `-Z mutable-noalias=no` or `-Z mutable-noalias=yes` is given.

<details><summary>Backtrace + Assembly</summary>
<p>

```
gdb) bt
#0  0x000055555555b77b in testmin::List::drain ()
#1  0x000055555555b8b2 in testmin::main ()
#2  0x000055555555b9b3 in std::sys_common::backtrace::__rust_begin_short_backtrace ()
#3  0x000055555555b6f9 in std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h0eb3c70533b77608 ()
#4  0x000055555556cd1a in core::ops::function::impls::{impl#2}::call_once<(), dyn core::ops::function::Fn<()>> ()
    at /rustc/7a16cfcffc58cd6994984a2c014b0cc27ed8f66b/library/core/src/ops/function.rs:259
#5  std::panicking::try::do_call<&dyn core::ops::function::Fn<()>, i32> () at library/std/src/panicking.rs:401
#6  std::panicking::try<i32, &dyn core::ops::function::Fn<()>> () at library/std/src/panicking.rs:365
#7  std::panic::catch_unwind<&dyn core::ops::function::Fn<()>, i32> () at library/std/src/panic.rs:434
#8  std::rt::lang_start_internal::{closure#2} () at library/std/src/rt.rs:45
#9  std::panicking::try::do_call<std::rt::lang_start_internal::{closure#2}, isize> () at library/std/src/panicking.rs:401
#10 std::panicking::try<isize, std::rt::lang_start_internal::{closure#2}> () at library/std/src/panicking.rs:365
#11 std::panic::catch_unwind<std::rt::lang_start_internal::{closure#2}, isize> () at library/std/src/panic.rs:434
#12 std::rt::lang_start_internal () at library/std/src/rt.rs:45
#13 0x000055555555b9a2 in main ()

(gdb) disass
Dump of assembler code for function _ZN7testmin4List5drain17hed4c20055222af0aE:
   0x000055555555b760 <+0>:	lea    0x10(%rdi),%rax
   0x000055555555b764 <+4>:	cmp    %rax,0x8(%rdi)
   0x000055555555b768 <+8>:	je     0x55555555b78b <_ZN7testmin4List5drain17hed4c20055222af0aE+43>
   0x000055555555b76a <+10>:	mov    (%rax),%rcx
   0x000055555555b76d <+13>:	nopl   (%rax)
   0x000055555555b770 <+16>:	mov    (%rcx),%rdx
   0x000055555555b773 <+19>:	mov    0x8(%rcx),%rsi
   0x000055555555b777 <+23>:	mov    %rsi,0x8(%rdx)
=> 0x000055555555b77b <+27>:	mov    (%rcx),%rdx
   0x000055555555b77e <+30>:	mov    0x8(%rcx),%rsi
   0x000055555555b782 <+34>:	mov    %rdx,(%rsi)
   0x000055555555b785 <+37>:	cmp    %rax,0x8(%rdi)
   0x000055555555b789 <+41>:	jne    0x55555555b770 <_ZN7testmin4List5drain17hed4c20055222af0aE+16>
   0x000055555555b78b <+43>:	ret    
```
</p>
Above, `0x10(%rdi)` contains `self.tail.prev` but it is not reloaded in the loop starting at 0x000055555555b770
</details>

If the behavior is due to my invoking UB, I would appreciate a pointer to where the UB lies.

I tried using MIRI which says:
<details><summary>MIRI Output</summary>

```
part 1, build a list with 2 elements
trying to drain list
error: Undefined Behavior: not granting access to tag <untagged> because incompatible item is protected: [SharedReadOnly for <6195> (call 1791)]
  --> src/bin/testmin.rs:22:13
   |
22 |             (*(*this).next).prev = (*this).prev;
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not granting access to tag <untagged> because incompatible item is protected: [SharedReadOnly for <6195> (call 1791)]
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
           
   = note: inside `ListElem::remove` at src/bin/testmin.rs:22:13
note: inside `List::pop_back` at src/bin/testmin.rs:48:9
  --> src/bin/testmin.rs:48:9
   |
48 |         ListElem::remove(self.tail.prev);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `List::drain` at src/bin/testmin.rs:54:13
  --> src/bin/testmin.rs:54:13
   |
54 |             self.pop_back();
   |             ^^^^^^^^^^^^^^^
note: inside `main` at src/bin/testmin.rs:81:5
  --> src/bin/testmin.rs:81:5
   |
81 |     l.drain();
   |     ^^^^^^^^^
```
which is an error I do not understand. From discussions on the users forum it was my impression that Rust's aliasing rules do not apply to raw pointers, but I may have misunderstood.

The code itself is derived from a standard sentinel-based doubly linked list.
</details>

**Comment by SkiFire13** [comment]
What you're doing is UB: `drain` is given a shared reference, so no code can legally mutate it while it is executing. This is what MIRI is complaining about.

**Comment by godmar** [comment]
> What you're doing is UB: drain is given a shared reference

Could you elaborate?  `drain` is given a shared reference to what?  `self`?

**Comment by SkiFire13** [comment]
Yes, I was refering to `self`

**Comment by godmar** [comment]
If I change `drain(&self)` to `drain(&mut self)` I observe the same behavior.

**Comment by SkiFire13** [comment]
Did you also change `pop_back`? The same problem applies to it

**Comment by godmar** [comment]
If I change both `drain` and `pop_back` to be `&mut self`, the MIRI warning remains the same, but the code completes.  When I make the same changes to the code from which the minimum test case was derived, however, the problem remains.  I'll try to produce a new test case in which I'll ensure that I have a mutable reference to `self` always.

**Comment by SkiFire13** [comment]
I think that even with those changes the code is still UB under the stacked borrows model because it observes `self` with a mutable reference and then modifies it through a pointer that wasn't derived from it.

**Comment by godmar** [comment]
Correction: with those changes, it works under `+stable` but still loops infinitely under `+nightly`

**Comment by godmar** [comment]
> I think that even with those changes the code is still UB under the stacked borrows model because it observes `self` with a mutable reference and then modifies it through a pointer that wasn't derived from it.

Does this mean that the pointer `self.tail.prev` is considered to be not derived from `self`?

**Comment by godmar** [comment]
Also, for the new playground (where `drain` and `pop_back` both use `&mut self`) the behavior depends on whether `-Z mutable-noalias=no` is given to nightly rustc or not.  If set to `no`, it'll complete. If set to `yes`, it'll loop.

**Comment by SkiFire13** [comment]
> Does this mean that the pointer self.tail.prev is considered to be not derived from self?

Yes because it is in fact not derived from the `self` inside `drain`, it is derived from an old reference to `self`, which is no longer valid when `drain` executes.

**Comment by godmar** [comment]
Could you explain more?
 `self.drain()` calls `self.pop_back()` which performs `self.tail.prev` - in what way is `self.tail.prev` derived from an "old reference to self?" that's "no longer valid when `drain` executes?"  The pointer isn't accessed until after `drain` has started executing (at least based on the way the code is written?) - why is the reference to `self` old at this point?

**Comment by godmar** [comment]
Are you referring to LLVM's ["based on"](https://llvm.org/docs/LangRef.html#pointer-aliasing-rules) relationship that is linked from the [Rust reference on undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#behavior-considered-undefined)?

It includes `getelementptr`, `bitcast` and `integerptr` which I think is a way of describing simple pointer arithmetic.  But it probably excludes pointers stored inside an object as would be the case here (`self.tail.prev` is a pointer stored inside `*self`).

**Comment by nikic** [comment]
Yes, that's correct. You can indirect a pointer through memory while preserving provenance, but as @SkiFire13 already said, it's an old reference to self that was stored, it's not derived from your current self.

Closing this as there does not seem to be any rustc bug here.

**Comment by godmar** [comment]
I think this explanation makes sense. The presence of a `&mut self` means that modifications via raw pointers to memory anywhere within the object, including embedded fields like `self.tail.prev` would be assumed to not be aliased. I'll have to analyze the consequences of this. Either forgo Rust references altogether, or perhaps research UnsafeCell some more.

This appears to be the case even if the access is via a raw pointer, as in:
```rust
    fn pop_back(&mut self) {
        let rawp : *const *mut ListElem = &mut self.tail.prev as *const *mut ListElem;
        unsafe {
            ListElem::remove(*rawp);
        }
    }
```
or
```rust
    fn pop_back(&mut self) {
        let rawp : *const ListElem = &mut self.tail as *const ListElem;
        unsafe {
            ListElem::remove((*rawp).prev);
        }
    }
```
or even
```rust
    fn pop_back(&mut self) {
        let rawself = self as *mut List;
        unsafe {
            let rawp : *mut ListElem = &mut (*rawself).tail as *mut ListElem;
            ListElem::remove((*rawp).prev);
        }
    }
```
all of which don't work.

**Comment by godmar** [comment]
> Yes, that's correct. You can indirect a pointer through memory while preserving provenance, but as @SkiFire13 already said, it's an old reference to self that was stored, it's not derived from your current self.
> 

Thanks. I think I misunderstood "current self" as the current value of the pointer representing the `self` reference, but I think what you really mean is the content of the thing to which `self` refers (akin to `*this` in C++).

**Comment by SkiFire13** [comment]
Sorry, I should have been more explicit. Even if the address of the pointer (the `&self`/`&mut self` that `drain` takes) and the value it points to are the same, what changes is the reference. For the compiler when you called `drain` you created a new reference, thus asserting unique access to `l`/`self`, which makes any raw pointer that pointed to it invalid.
